src/frontend/src/optimizer/rule/apply_proj_agg.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_expr::expr::AggKind;
+
+use super::{BoxedRule, Rule};
+use crate::expr::{ExprImpl, ExprRewriter, InputRef};
+use crate::optimizer::plan_node::{
+    LogicalAgg, LogicalProject, PlanTreeNodeBinary, PlanTreeNodeUnary,
+};
+use crate::optimizer::PlanRef;
+use crate::utils::ColIndexMapping;
+
+/// This rule is for pattern: Apply->Project->Agg, and it will be converted into Project->Agg->Apply.
+/// Scalar agg will be converted into group agg using all columns of apply's left child, which is different from original formula.
+/// Project will have all columns of Apply's left child at its beginning.
+/// Please note that Project's input is Agg here, so we don't have to worry about edge cases of pulling up Project.
+pub struct ApplyProjAggRule {}
```

Comment:
How about naming it as `AggUnnestRule`? One cannot get what it is from `ApplyProjAggRule`

Reply:
BTW, others may not know what is "original formula". Perhaps you may just explain the idea of pulling Agg across Apply from scratch

Reply:
BTW*2, Please add an empty line between paragraphs, which is required by the Markdown render

Reply:
ApplyProjAgg matches the pattern: Apply->Proj->Agg........
This rule doesn't handle the general case of unnesting agg, so AggUnnestRule doesn't look that accurate?

Reply:
Yes, that describes the match pattern, but I think it's more meaningful to name it with what it does.  e.g. `ProjectMergeRule` is a good name, but `FilterJoinRule` is not so good, perhaps `FilterJoinTransposeRule` sounds better

Reply:
How about PullUpCorrelatedPredicate?